### PR TITLE
allow to generate a pex executable

### PR DESCRIPTION
**Pull request by sileht**
allow to generate a pex executable

Howto use:
```
    $ tox -epy35-pex
    $ dist/gnocchi-4.2.1.dev69-py35-linux_x86_64 config-generator > gnocchi.conf
    $ dist/gnocchi-4.2.1.dev69-py35-linux_x86_64 api --config-file gnocchi.conf

```

### CODE REVIEW

**Comment by sileht** [comment]
Pull-request updated, HEAD is now a920e08ae1783a3260d25d4bfcf06852a8f40dc0

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 80ffaadafa71559f0a11c48d9f09d87bf9895d37

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 5e21cd5b5058406d8cf75c181a56165c34032c3e

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 17a0b060fdbd8e34feb3a08b705893f08cd4eb6d

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 17a0b060fdbd8e34feb3a08b705893f08cd4eb6d

**Comment by sileht** [comment]
@chungg The idea is to ship an all in one binary like people does with Go

To do so, the idea is to use the capability of python to execute __main__.py of a zip.

pex is a lib developed by twitter to create a zip that contains the wheels of the applications and all its deps, and it creates a __main__.py that exec the application. On top of the zip a sheband is added (#/usr/bin/env python). So you can just run the zip like any other binary

**Comment by chungg** [comment]
fancy :)

File: tox.ini [hunk 1]
```diff
@@ -127,6 +127,16 @@ exclude = .tox,.eggs,doc,gnocchi/rest/prometheus/remote_pb2.py
 show-source = true
 enable-extensions = H904
 
+[testenv:py35-pex]
+deps =
+  pex
+  requests
```
Comment (jd) [inline tox.ini]: what is this for?

File: tox.ini [hunk 2]
```diff
@@ -127,6 +127,16 @@ exclude = .tox,.eggs,doc,gnocchi/rest/prometheus/remote_pb2.py
 show-source = true
 enable-extensions = H904
 
+[testenv:py35-pex]
```
Comment (jd) [inline tox.ini]: just go for a py3-pex and use a `basepython = python3` maybe? no need to be specific.

File: tox.ini [hunk 3]
```diff
@@ -127,6 +127,16 @@ exclude = .tox,.eggs,doc,gnocchi/rest/prometheus/remote_pb2.py
...
+deps =
+  pex
+  requests
+commands = {toxinidir}/tools/build_pex.sh {envname}
+
+[testenv:py27-pex]
```
Comment (jd) [inline tox.ini]: ditto

File: tools/build_pex.sh [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+# * setproctitle-1.1.10-cp35-cp35m-linux_x86_64.whl
+# * SQLAlchemy-1.2.5-cp35-cp35m-linux_x86_64.whl
+# * ujson-1.35-cp35-cp35m-linux_x86_64.whl
+# * wrapt-1.10.11-cp35-cp35m-linux_x86_64.whl
+EXTRAS=mysql,postgresql,redis,keystone,swift,s3,ceph
+
+BINARY=gnocchi-${VERSION}-${PYVERS}-${OS}_${ARCH}
+
+mkdir -p dist
+rm $CACHE_DIR/gnocchi-* -rf
```
Comment (jd) [inline tools/build_pex.sh]: `rm -rf $CACHE_DIR/gnocchi-*`

File: gnocchi/pex.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
+#
```
Comment (jd) [inline gnocchi/pex.py]: !

**Comment by jd** [comment]
Can we run this on gnocchi.xyz to publish Linux binaries?

File: tox.ini (hunk 2 above)
Comment (sileht) [reply tox.ini]: I changed that but the binary generated will still be specific.

File: tools/build_pex.sh (hunk 4 above)
Comment (sileht) [reply tools/build_pex.sh]: done

File: gnocchi/pex.py (hunk 5 above)
Comment (sileht) [reply gnocchi/pex.py]: done

File: tox.ini (hunk 1 above)
Comment (sileht) [reply tox.ini]: I have used pex extras instead.

**Comment by jd** [comment]
@sileht I guess you forgot to push.

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 083b48f0a07b9abb97dee0f19b7153404c8418d4

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 7fc50bd3ad3ab839b99f8ed27d491f416211943c

File: doc/source/pex.j2 [hunk 6]
```diff
@@ -0,0 +1,30 @@
+Installation from pex binary
+============================
+
+Gnocchi can be install by just downloading a binary (built with `pex`_).
```
Comment (jd) [inline doc/source/pex.j2]: installed

File: doc/source/pex.j2 [hunk 7]
```diff
@@ -0,0 +1,30 @@
+Installation from pex binary
```
Comment (jd) [inline doc/source/pex.j2]: Could this file by named `pex.rst.j2` ?

File: doc/source/pex.j2 [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    lrwxrwxrwx 1 user user       37 Mar 20 07:53 gnocchi-api -> {{ master_binary }}
+    lrwxrwxrwx 1 user user       37 Mar 20 07:53 gnocchi-change-sack-size -> {{ master_binary }}
+    lrwxrwxrwx 1 user user       37 Mar 20 07:53 gnocchi-config-generator -> {{ master_binary }}
+    lrwxrwxrwx 1 user user       37 Mar 20 07:53 gnocchi-metricd -> {{ master_binary }}
+    lrwxrwxrwx 1 user user       37 Mar 20 07:53 gnocchi-statsd -> {{ master_binary }}
+    lrwxrwxrwx 1 user user       37 Mar 20 07:53 gnocchi-upgrade -> {{ master_binary }}
+
+.. _pex: https://github.com/pantsbuild/pex
+
+
```
Comment (jd) [inline doc/source/pex.j2]: empty lines

File: doc/source/pex.j2 [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    $ ls -l
+    total 33036
+    -rwxr-xr-x 1 user user 33828159 Mar 20 07:52 {{ master_binary }}
+    lrwxrwxrwx 1 user user       37 Mar 20 07:53 gnocchi-api -> {{ master_binary }}
+    lrwxrwxrwx 1 user user       37 Mar 20 07:53 gnocchi-change-sack-size -> {{ master_binary }}
+    lrwxrwxrwx 1 user user       37 Mar 20 07:53 gnocchi-config-generator -> {{ master_binary }}
+    lrwxrwxrwx 1 user user       37 Mar 20 07:53 gnocchi-metricd -> {{ master_binary }}
+    lrwxrwxrwx 1 user user       37 Mar 20 07:53 gnocchi-statsd -> {{ master_binary }}
+    lrwxrwxrwx 1 user user       37 Mar 20 07:53 gnocchi-upgrade -> {{ master_binary }}
+
```
Comment (jd) [inline doc/source/pex.j2]: Add a paragraph here like:

> You could then use run those symbolic links as they were standard executable. For example, call `./gnocchi-upgrade` to run `gnocchi-upgrade`.

File: doc/source/pex.j2 (hunk 6 above)
Comment (sileht) [reply doc/source/pex.j2]: done

File: doc/source/pex.j2 (hunk 7 above)
Comment (sileht) [reply doc/source/pex.j2]: done

File: doc/source/pex.j2 (hunk 9 above)
Comment (sileht) [reply doc/source/pex.j2]: done

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 5e08ac135f8437ec433edbf64f54c86c9997a111

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 820665a3701b93c786469c5cc181cc94f21c305a

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 7d4c1fb8ea4e6ca63893cf5d9f0c8b46a471e4b2

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 84c19b3515ac421925233073f5ae2d8b20fc603f

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 1fc35a5a179ca32db6854737b500c479f89581b7

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 737fa0a3447cea2c666f327bc0ddd3fbc60f91ca

**Pull request closed without merging**